cmd/argocd/commands/app.go

```diff
@@ -336,6 +337,62 @@ func NewApplicationListCommand(clientOpts *argocdclient.ClientOptions) *cobra.Co
 	return command
 }
 
+// NewApplicationWaitCommand returns a new instance of an `argocd app wait` command
+func NewApplicationWaitCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
+	var (
+		syncOnly   bool
+		healthOnly bool
+		timeout    uint
+	)
+	const (
+		DEFAULT_CHECK_TIMEOUT_SECONDS  = 10
+		DEFAULT_CHECK_INTERVAL_SECONDS = 5
+	)
+	var command = &cobra.Command{
+		Use:   "wait APPNAME",
+		Short: "Wait for application to reach a target state",
+		Run: func(c *cobra.Command, args []string) {
+			if len(args) != 1 {
+				c.HelpFunc()(c, args)
+				os.Exit(1)
+			}
+			if syncOnly && healthOnly {
+				log.Fatalln("Please specify at most one of --sync-only or --health-only.")
+			}
+			conn, appIf := argocdclient.NewClientOrDie(clientOpts).NewApplicationClientOrDie()
+			defer util.Close(conn)
+
+			appName := args[0]
+
+			success := util.Wait(DEFAULT_CHECK_INTERVAL_SECONDS, timeout, func() bool {
+				app, err := appIf.Get(context.Background(), &application.ApplicationQuery{Name: &appName})
+				if err != nil {
+					log.Fatal(err)
+				}
+
+				log.Printf("Checking app %q...", appName)
+				if !healthOnly && app.Status.ComparisonResult.Status != argoappv1.ComparisonStatusSynced {
+					return false
+				}
+				if !syncOnly && app.Status.Health.Status != argoappv1.HealthStatusHealthy {
+					return false
+				}
+
+				return true
+			})
+			if success {
+				log.Printf("App %q matches desired state", appName)
+			} else {
+				log.Printf("Timed out before seeing app %q match desired state", appName)
```

Comment:
Please print more details to stderr if app has not reached desired state:

if app is not healthy then we need to print list of unhealthy components and reason
if app is not synced then we need to list of unsynced components